tests/unit/steps/test_steps_analyze.py

```diff
@@ -55,31 +55,73 @@ async def test_analysis_dump(mocker, apps):
         "    cinder/2:\n"
         "      workload_version: 16.4.2\n"
         "      os_version: ussuri\n"
+        "\n"
+        "rabbitmq-server:\n"
+        "  model_name: my_model\n"
+        "  charm: rabbitmq-server\n"
+        "  charm_origin: ch\n"
+        "  os_origin: distro\n"
+        "  channel: 3.8/stable\n"
+        "  units:\n"
+        "    rabbitmq-server/0:\n"
+        "      workload_version: '3.8'\n"
+        "      os_version: yoga\n"
     )
 
-    mocker.patch.object(analyze.Analysis, "_populate", return_value=apps)
+    mocker.patch.object(
+        analyze.Analysis,
+        "_populate",
+        return_value=[apps["keystone_ussuri"], apps["cinder_ussuri"], apps["rmq_ussuri"]],
+    )
     result = await analyze.Analysis.create()
     assert str(result) == expected_result
 
 
 @pytest.mark.asyncio
-async def test_generate_model(mocker, full_status, config):
+async def test_populate_model(mocker, full_status, config):
     mocker.patch.object(analyze, "async_get_status", return_value=full_status)
     mocker.patch.object(
         analyze, "async_get_application_config", return_value=config["openstack_ussuri"]
     )
-    # Initially, 3 applications are in the status (keystone, cinder and rabbitmq-server)
-    assert len(full_status.applications) == 3
+    # Initially, 4 applications are in the status: keystone, cinder, rabbitmq-server and my-app
+    # my-app is a unknown application in the model.
+    assert len(full_status.applications) == 4
     apps = await Analysis._populate()
-    assert len(apps) == 3
-    assert {app.charm for app in apps} == {"keystone", "cinder", "rabbitmq-server"}
+    assert len(apps) == 4
+    # apps are on the UPGRADE_ORDER sequence and unknown in the end of the list
+    assert [app.charm for app in apps] == ["rabbitmq-server", "keystone", "cinder", "my-app"]
 
 
 @pytest.mark.asyncio
 async def test_analysis(mocker, apps):
```

Comment:
This test packs a lot of logic! Ideally, tests should be as self-evident as possible, contain the least amount of logic possible, only contain one assertion per test (ideally), and be unashamedly verbose if needed. This is so that when something fails, you can more quickly understand what did.

iiuc you are testing that:
* the Analysis object is built correctly
* that the current_cloud_os_release property reacts correctly to three separate cases:
  * subordinates are skipped
  * non-OpenStack charms are skipped
  * OpenStack charms on the same version yield the correct result

I suggest splitting this test in at least three tests (or one with parametrization), focusing on `current_cloud_os_release`, and adding a case including OpenStack charms of different versions (to test the min algorithm).

Reply:
Good catch. Thanks!
I've splitted the unit tests to separate what we are testing.